iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/resource/tsfile/PipeTsFileResource.java

```diff
@@ -180,35 +154,63 @@ synchronized boolean tryCacheObjects() throws IOException {
+    // is hard to shrink and may consume too much memory.
     allocatedMemoryBlock =
         PipeResourceManager.memory()
             .forceAllocateIfSufficient(
                 PipeConfig.getInstance().getPipeMemoryAllocateForTsFileSequenceReaderInBytes(),
-                0.5f);
+                MEMORY_SUFFICIENT_THRESHOLD);
     if (allocatedMemoryBlock == null) {
+      LOGGER.info(
+          "PipeTsFileResource: Failed to create TsFileSequenceReader for tsfile {} in cache, because memory usage is high",
+          hardlinkOrCopiedFile.getPath());
       return false;
     }
+    allocatedMemoryBlock.close();
+    allocatedMemoryBlock = null;
 
-    LOGGER.info(
-        "PipeTsFileResource: Cached objects for tsfile {}.", hardlinkOrCopiedFile.getPath());
+    long memoryRequiredInBytes = 0L;
+    try (TsFileSequenceReader sequenceReader =
+        new TsFileSequenceReader(hardlinkOrCopiedFile.getPath(), true, true)) {
+      deviceMeasurementsMap = sequenceReader.getDeviceMeasurementsMap();
+      memoryRequiredInBytes += PipeMemoryWeighUtil.memoryOfStr2StrList(deviceMeasurementsMap);
+
+      deviceIsAlignedMap = new HashMap<>();
+      final TsFileDeviceIterator deviceIsAlignedIterator =
+          sequenceReader.getAllDevicesIteratorWithIsAligned();
+      while (deviceIsAlignedIterator.hasNext()) {
+        final Pair<String, Boolean> deviceIsAlignedPair = deviceIsAlignedIterator.next();
+        deviceIsAlignedMap.put(deviceIsAlignedPair.getLeft(), deviceIsAlignedPair.getRight());
+      }
+      memoryRequiredInBytes += PipeMemoryWeighUtil.memoryOfStr2Bool(deviceIsAlignedMap);
 
-    sequenceReader = new TsFileSequenceReader(hardlinkOrCopiedFile.getPath(), true, true);
-    reader = new TsFileReader(sequenceReader);
-    deviceMeasurementsMap = sequenceReader.getDeviceMeasurementsMap();
-    deviceIsAlignedMap = new HashMap<>();
-    final TsFileDeviceIterator deviceIsAlignedIterator =
-        sequenceReader.getAllDevicesIteratorWithIsAligned();
-    while (deviceIsAlignedIterator.hasNext()) {
-      final Pair<String, Boolean> deviceIsAlignedPair = deviceIsAlignedIterator.next();
-      deviceIsAlignedMap.put(deviceIsAlignedPair.getLeft(), deviceIsAlignedPair.getRight());
+      measurementDataTypeMap = sequenceReader.getFullPathDataTypeMap();
+      memoryRequiredInBytes += PipeMemoryWeighUtil.memoryOfStr2TSDataType(measurementDataTypeMap);
     }
-    measurementDataTypeMap = sequenceReader.getFullPathDataTypeMap();
 
-    // Clear cached device metadata to save memory.
-    sequenceReader.clearCachedDeviceMetadata();
+    // Allocate again for the cached objects.
+    allocatedMemoryBlock =
+        PipeResourceManager.memory()
+            .forceAllocateIfSufficient(memoryRequiredInBytes, MEMORY_SUFFICIENT_THRESHOLD);
+    if (allocatedMemoryBlock == null) {
+      LOGGER.info(
+          "PipeTsFileResource: Failed to cache objects for tsfile {} in cache, because memory usage is high",
```

Comment:
The "PipeTsFileResource" may already contained in "LOGGER".

Reply:
@SteveYurongSu